soc/riscv/riscv-ite/it8xxx2/soc.c

```diff
@@ -9,6 +9,133 @@
+	/*
+	 * PLL frequency setting = 4 (48MHz)
+	 * FND   div = 0 (PLL / 1 = 48 mhz)
+	 * UART  div = 1 (PLL / 2 = 24 mhz)
+	 * SMB   div = 1 (PLL / 2 = 24 mhz)
+	 * SSPI  div = 1 (PLL / 2 = 24 mhz)
+	 * EC    div = 6 (FND / 6 =  8 mhz)
+	 * JTAG  div = 1 (PLL / 2 = 24 mhz)
+	 * PWM   div = 0 (PLL / 1 = 48 mhz)
+	 * USBPD div = 5 (PLL / 6 =  8 mhz)
+	 */
+	{.pll_freq  = 4,
+	 .div_fnd   = 0,
+	 .div_uart  = 1,
+	 .div_smb   = 1,
+	 .div_sspi  = 1,
+	 .div_ec    = 6,
+	 .div_jtag  = 1,
+	 .div_pwm   = 0,
+	 .div_usbpd = 5}
+};
+
+void __intc_ram_code chip_pll_ctrl(enum chip_pll_mode mode)
+{
+	IT8XXX2_ECPM_PLLCTRL = mode;
+	/* for deep doze / sleep mode */
+	IT8XXX2_ECPM_PLLCTRL = mode;
+}
+
+void __intc_ram_code chip_run_pll_sequence(const struct pll_config_t *pll)
+{
+	/* Enable HW timer to wakeup chip from the sleep mode */
+	timer_5ms_one_shot();
+	/*
+	 * Configure PLL clock dividers.
+	 * Writing data to these registers doesn't change the
+	 * PLL frequency immediately until the status is changed
+	 * into wakeup from the sleep mode.
+	 * The following code is intended to make the system
+	 * enter sleep mode, and wait HW timer to wakeup chip to
+	 * complete PLL update.
+	 */
+	IT8XXX2_ECPM_PLLFREQR = pll->pll_freq;
+	/* Pre-set FND clock frequency = PLL / 3 */
+	IT8XXX2_ECPM_SCDCR0 = (2 << 4);
+	/* JTAG and EC */
+	IT8XXX2_ECPM_SCDCR3 = (pll->div_jtag << 4) | pll->div_ec;
+	/* Chip sleep after wait for interrupt (wfi) instruction */
+	chip_pll_ctrl(CHIP_PLL_SLEEP);
+	/* Chip sleep and wait timer wake it up */
+	__asm__ volatile  ("wfi");
+	/* New FND clock frequency */
+	IT8XXX2_ECPM_SCDCR0 = pll->div_fnd << 4;
+	/* Chip doze after wfi instruction */
+	chip_pll_ctrl(CHIP_PLL_DOZE);
+	/* UART */
+	IT8XXX2_ECPM_SCDCR1 = pll->div_uart;
+	/* SSPI and SMB */
+	IT8XXX2_ECPM_SCDCR2 = (pll->div_sspi << 4) | pll->div_smb;
+	/* USBPD and PWM */
+	IT8XXX2_ECPM_SCDCR4 = (pll->div_usbpd << 4) | pll->div_pwm;
+}
+
+static void chip_configure_pll(const struct pll_config_t *pll)
+{
+	/* Re-configure PLL clock or not. */
+	if (((IT8XXX2_ECPM_PLLFREQR & 0xf) != pll->pll_freq) ||
+		((IT8XXX2_ECPM_SCDCR0 & 0xf0) != (pll->div_fnd << 4)) ||
+		((IT8XXX2_ECPM_SCDCR3 & 0xf) != pll->div_ec)) {
+#ifdef CONFIG_ESPI
+		/*
+		 * TODO: implement me
+		 * We have to disable eSPI pad before changing
+		 * PLL sequence or sequence will fail if CS# pin is low.
+		 */
+#endif
+		/* Run change PLL sequence */
+		chip_run_pll_sequence(pll);
+#ifdef CONFIG_ESPI
+		/*
+		 * TODO: implement me
+		 * Enable eSPI pad after changing PLL sequence
+		 */
+#endif
+	}
+}
+
+#define IT8XXX2_IER_COUNT (IT8XXX2_IRQ_COUNT / 8 + 1)
+static uint8_t ier_setting[IT8XXX2_IER_COUNT];
+static int chip_change_pll(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	/* store enabled interrupts of chip-level */
+	for (int i = 0; i < IT8XXX2_IER_COUNT; i++)
+		ier_setting[i] = *reg_enable[i];
+	/* disable all interrupts */
+	for (int i = 0; i < IT8XXX2_IER_COUNT; i++)
+		*reg_enable[i] = 0;
```

Comment:
Can you create routines in the INTC driver for saving and restoring the interrupt state instead of accessing the INTC registers here?

ite_intc_save_and_disable_interrupts()
ite_intc_restore_interrupts()

Also calls to these functions should be guarded with IS_ENABLED(CONFIG_ITE_IT8XXX2_INTC).

Reply:
If these are moved into INTC, which I like, how much protection needs to be taken for multiple saves before multiple restores?  If it is left dangerous, make sure a comment is added in the function header to indicate that it does not allow for nesting.

Reply:
Done.